lark/tree.py

```diff
@@ -3,19 +3,54 @@
 except ImportError:
     pass
 
+from collections import deque
 from copy import deepcopy
 
 from .utils import inline_args
 
 ###{standalone
 class Tree(object):
-    def __init__(self, data, children):
+    __slots__ = ('data', 'children', 'tags')
+
+    def __init__(self, data, children, tags=None):
         self.data = data
         self.children = children
+        self.tags = tags
 
     def __repr__(self):
         return 'Tree(%s, %s)' % (self.data, self.children)
 
+    def __iter__(self):
+        return iter(self.children)
+
+    def __len__(self):
+        return len(self.children)
+
+    def __hasitem__(self, tag):
+        return self.tags and tag in self.tags
+
+    def __getitem__(self, tag):
+        if self.tags is None:
+            raise KeyError(tag)
+        return self.tags[tag]
+
+    def __setitem__(self, tag, value):
+        if self.tags is None:
+            self.tags = {}
+        self.tags[tag] = value
```

Comment:
Since it uses `__slots__` we can't attach arbitrary attributes to an instance. This mechanism solves the problem by allowing parsers/transformers to tag the node using slicing.